bpf/mb_xdp.c

```diff
@@ -0,0 +1,212 @@
+        return XDP_PASS;
+    }
+
+    struct iphdr *iph = (struct iphdr *)(eth + 1);
+    if ((void *)(iph + 1) > data_end) {
+        return XDP_DROP;
+    }
+
+    if (iph->protocol == IPPROTO_IPIP) {
+        iph = ((void *)iph + iph->ihl * 4);
+        if ((void *)(iph + 1) > data_end) {
+            return XDP_PASS;
+        }
+    }
+
+    if (iph->protocol != IPPROTO_TCP) {
+        return XDP_PASS;
+    }
+
+    struct tcphdr *tcph = (struct tcphdr *)(iph + 1);
+    if ((void *)(tcph + 1) > data_end) {
+        return XDP_DROP;
+    }
+    if (tcph->syn && !tcph->ack) {
+        // first packet
+        if (bpf_htons(tcph->dest) == IN_REDIRECT_PORT) {
+            // same node, already rewrite dest port by connect.
+            // bypass.
+            return XDP_PASS;
+        }
+        // ingress without mb_connect
+        __u32 ip = iph->daddr;
+        struct pod_config *pod = bpf_map_lookup_elem(&local_pod_ips, &ip);
+        if (!pod) {
+            // dest ip is not on this node or not injected sidecar.
+            return XDP_PASS;
+        }
+        if (bpf_htons(tcph->dest) == pod->status_port) {
+            return XDP_PASS;
+        }
+        if (pod->exclude_in_ports[0] != 0) {
+            for (int i = 0; i < MAX_ITEM_LEN; i++) {
+                if (bpf_htons(tcph->dest) == pod->exclude_in_ports[i]) {
+                    return XDP_PASS;
+                }
+            }
+        }
+        int find = 0;
+        if (pod->include_in_ports[0] != 0) {
+            for (int i = 0; i < MAX_ITEM_LEN; i++) {
+                if (bpf_htons(tcph->dest) == pod->include_in_ports[i]) {
+                    find = 1;
+                    break;
+                }
+            }
+        } else {
+            find = 1;
+        }
+        if (!find) {
+            return XDP_PASS;
+        }
+
+        // like from 10.0.0.1:23456 => 172.31.0.123:80
+        // we will rewrite the dest port from 80 to 15006
+        // which will be: 10.0.0.1:23456 => 172.31.0.123:15006
+        struct pair p = {
+            .sip = iph->saddr,
+            .sport = tcph->source,
+            .dip = iph->daddr,
+            .dport = bpf_htons(IN_REDIRECT_PORT),
+        };
+        struct origin_info origin = {
+            .ip = iph->daddr,
+            .port = tcph->dest,
+            .flags = XDP_ORIGIN_FLAG,
+        };
+        bpf_map_update_elem(&pair_original_dst, &p, &origin, BPF_NOEXIST);
+        __u16 oldd = tcph->dest;
+        tcph->dest = bpf_htons(IN_REDIRECT_PORT); // rewrite dest port.
+        tcph->check = csum_diff4(oldd, tcph->dest, tcph->check);
+        return XDP_PASS;
+    } else if (tcph->fin && tcph->ack) {
+        // todo delete key
+        struct pair p = {
+            .dip = iph->saddr,
+            .dport = tcph->source,
+            .sip = iph->daddr,
+            .sport = bpf_htons(IN_REDIRECT_PORT),
+        };
+        bpf_map_delete_elem(&pair_original_dst, &p);
+        return XDP_PASS;
```

Comment:
Should we rewrite dst/src port when client/server closing the connection?

Reply:
request and response logic have covered this, we do not need do this manually.